compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -1822,20 +1822,17 @@ object TypeComparer {
 
   /** Show trace of comparison operations when performing `op` as result string */
   def explaining[T](say: String => Unit)(op: Context => T)(implicit ctx: Context): T = {
-    val (res, explanation) = underlyingExplained(op)
-    say(explanation)
+    val nestedCtx = ctx.fresh.setTypeComparerFn(new ExplainingTypeComparer(_))
+    val res = op(nestedCtx)
+    say(nestedCtx.typeComparer.lastTrace())
     res
   }
 
   /** Like [[explaining]], but returns the trace instead */
   def explained[T](op: Context => T)(implicit ctx: Context): String = {
-    underlyingExplained(op)._2
-  }
-
-  private def underlyingExplained[T](op: Context => T)(implicit ctx: Context): (T, String) = {
-    val nestedCtx = ctx.fresh.setTypeComparerFn(new ExplainingTypeComparer(_))
-    val res = op(nestedCtx)
-    (res, nestedCtx.typeComparer.lastTrace())
+    var trace: String = null
+    explaining(trace = _)(op)
+    trace
```

Comment:
I tried to adjust for https://github.com/lampepfl/dotty/pull/5611#discussion_r241851303 by basing `.explained` on `.explaining`. Basing it the other way around doesn't work, since we can't assing `null` to `var res: T`.

I replaced `.toString()` on `ExplainingTypeComparer` with `.lastTrace()`, because the previous implementation interacted badly with debugging. Every time I stopped in TypeComparer, its `.toString` was called by the debugger, which got rid of the stack trace.